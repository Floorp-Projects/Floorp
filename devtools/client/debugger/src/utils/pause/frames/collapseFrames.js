/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { findIndex } from "lodash";

// eslint-disable-next-line max-len
import { getFrameUrl } from "./getFrameUrl";

function collapseLastFrames(frames) {
  const index = findIndex(frames, frame =>
    getFrameUrl(frame).match(/webpack\/bootstrap/i)
  );

  if (index == -1) {
    return { newFrames: frames, lastGroup: [] };
  }

  const newFrames = frames.slice(0, index);
  const lastGroup = frames.slice(index);
  return { newFrames, lastGroup };
}

export function collapseFrames(frames) {
  // We collapse groups of one so that user frames
  // are not in a group of one
  function addGroupToList(group, list) {
    if (!group) {
      return list;
    }

    if (group.length > 1) {
      list.push(group);
    } else {
      list = list.concat(group);
    }

    return list;
  }
  const { newFrames, lastGroup } = collapseLastFrames(frames);
  frames = newFrames;
  let items = [];
  let currentGroup = null;
  let prevItem = null;
  for (const frame of frames) {
    const prevLibrary = prevItem?.library;

    if (!currentGroup) {
      currentGroup = [frame];
    } else if (prevLibrary && prevLibrary == frame.library) {
      currentGroup.push(frame);
    } else {
      items = addGroupToList(currentGroup, items);
      currentGroup = [frame];
    }

    prevItem = frame;
  }

  items = addGroupToList(currentGroup, items);
  items = addGroupToList(lastGroup, items);
  return items;
}
