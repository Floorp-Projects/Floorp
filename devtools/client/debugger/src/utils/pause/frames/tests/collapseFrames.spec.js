/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { collapseFrames } from "../collapseFrames";

describe("collapseFrames", () => {
  it("default", () => {
    const groups = collapseFrames([
      { displayName: "a" },

      { displayName: "b", library: "React" },
      { displayName: "c", library: "React" },
    ]);

    expect(groups).toMatchSnapshot();
  });

  it("promises", () => {
    const groups = collapseFrames([
      { displayName: "a" },

      { displayName: "b", library: "React" },
      { displayName: "c", library: "React" },
      {
        displayName: "d",
        library: undefined,
        asyncCause: "promise callback",
      },
      { displayName: "e", library: "React" },
      { displayName: "f", library: "React" },
      { displayName: "g", library: undefined, asyncCause: null },
    ]);

    expect(groups).toMatchSnapshot();
  });
});
