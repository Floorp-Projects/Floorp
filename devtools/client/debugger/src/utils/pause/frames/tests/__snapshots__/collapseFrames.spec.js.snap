// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapseFrames default 1`] = `
Array [
  Object {
    "displayName": "a",
  },
  Array [
    Object {
      "displayName": "b",
      "library": "React",
    },
    Object {
      "displayName": "c",
      "library": "React",
    },
  ],
]
`;

exports[`collapseFrames promises 1`] = `
Array [
  Object {
    "displayName": "a",
  },
  Array [
    Object {
      "displayName": "b",
      "library": "React",
    },
    Object {
      "displayName": "c",
      "library": "React",
    },
  ],
  Object {
    "asyncCause": "promise callback",
    "displayName": "d",
    "library": undefined,
  },
  Array [
    Object {
      "displayName": "e",
      "library": "React",
    },
    Object {
      "displayName": "f",
      "library": "React",
    },
  ],
  Object {
    "asyncCause": null,
    "displayName": "g",
    "library": undefined,
  },
]
`;
